packages/cw1155/src/msg.rs

```diff
@@ -0,0 +1,48 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::{Binary, HumanAddr, Uint128};
+
+pub type TokenId = String;
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum Cw1155HandleMsg {
+    /// TransferFrom is a base message to move tokens, if `env.sender` has sufficient pre-approval.
+    TransferFrom {
+        // `None` means minting
+        from: Option<HumanAddr>,
+        // `None` means burning
+        to: Option<HumanAddr>,
+        token_id: TokenId,
+        value: Uint128,
+    },
+    /// SendFrom is a base message to move tokens to contract
+    /// if `env.sender` has sufficient pre-approval.
+    SendFrom {
+        // `None` means minting
+        from: Option<HumanAddr>,
+        contract: HumanAddr,
+        token_id: TokenId,
+        value: Uint128,
+        msg: Option<Binary>,
+    },
+    /// BatchTransferFrom is a base message to move tokens to another account without triggering actions
+    BatchTransferFrom {
+        // `None` means minting
+        from: Option<HumanAddr>,
+        // `None` means burning
+        to: Option<HumanAddr>,
+        batch: Vec<(TokenId, Uint128)>,
+    },
+    /// BatchSendFrom is a base message to move tokens to another to without triggering actions
+    BatchSendFrom {
+        // `None` means minting
+        from: Option<HumanAddr>,
+        contract: HumanAddr,
+        batch: Vec<(TokenId, Uint128)>,
+        msg: Option<Binary>,
+    },
+    /// Enable or disable approval for a third party ("operator") to manage all of the caller's tokens.
+    SetApprovalForAll { operator: HumanAddr, approved: bool },
```

Comment:
In cw721 we took a few liberties to "improve" the Ethereum design, even with the "all or nothing" approval. How about adding an [optional expiration date](https://github.com/CosmWasm/cosmwasm-plus/blob/master/packages/cw721/src/msg.rs#L38)?

Adding approvals on token by token basis would be nice as well, even if it is full control of the balance on that token_id. Are you concerned about performance? Or prefer the API this way? Especially iteration is MUCH more efficient in CosmWasm/Cosmos SDK than Ethereum and it is nice to embrace it.

Reply:
I think the only trade-off is use cases vs simplicity and on-chain resource usage.
For example, for the [example cw721 contract](https://github.com/CosmWasm/cosmwasm-plus/blob/master/contracts/cw721-base/src/state.rs#L13) to support approval by token, it needs to store a `Vec<Approval>` in `TokenInfo`, and need to do `O(n)` search when checking permission.

Reply:
Fair enough. This does add more runtime complexity. We can leave it as is.